paper.tex

```diff
@@ -329,6 +306,114 @@ \subsection{Unrolled recursion}
+% TODO: finish this \begin{algorithm} snippet
+\begin{algorithm}
+  $tid \gets$ global invocation ID\;
+  $(style, segment) \gets readScene(tid)$\;
+  \eIf{$style$ is $fill$}{
+      flatten source curve
+  }{\If{you're sad}{
+      just keep going
+  }}
+\caption{The control flow of our compute shader}
+\end{algorithm}
+
+Each thread outputs a polyline that fits a subset of the rendered subpath's outer contour, represented by the path segment. For a filled primitive, a thread only outputs the flattened approximation of the path segment. For a stroke, a thread outputs the polylines approximating multiple curves: a) the two parallel curves on both sides of the source segment, offset by the desired stroke \emph{halfwidth}, and b) the inner and outer joins that connect the path segment to the next adjacent segment. We also store metadata in our input encoding that marks the position of a path segment within the containing subpath. We use this information to determine whether to output an end cap instead of a join. We encode one additional segment designated as the \emph{stroke cap marker} for every subpath. The thread assigned to the marker segment only outputs a start cap, which completes the stroke outline.
+
+The joins between adjacent segments need to form a continuous outline when combined with the segments. This requires that a thread have access to the tangent vector at the end of its assigned segment \emph{and} the start tangent of the next adjacent segment. We require that the input segments within a subpath are stored in order which simplifies the access to the adjacent segment down to a simple read at the next array offset. Co-operation across threads is not necessary.
+
+A property of this data-parallel structure is that the output is unordered. The cubic-Bézier-to-Euler-spiral conversion is sequential within a single thread, so it is possible to preserve ordering at the scope of an individual path segment, if desired. However, ordering cannot be guaranteed across path segments. We refer to this output data structure as \emph{line soup}. We believe this structure can be adapted to other primitives, such as circular arc segments or triangles. We incorporated our compute shader into our own massively-parallel GPU rasterizer that spatially sorts the line soup in subsequent binning and tiling stages. As such, maintaining any sort of ordering with respect to the input path data structure was not necessary.
+
+% TODO: Talk about the inner join here. We detect inner vs outer outline based on join tangent cross product. We just draw a straight line for the inner join.
+% TODO: line soup is bump allocated. Mention this here? I'm leaning towards discussing that in the Limitations section below, which will focus on memory allocation and estimation.
+% TODO: while parallel on input, sequential per curve. However iterations are bounded (? needs measurement). Should this discussion happen here or down in the Performance section?
```

Comment:
I don't have strong feelings on where the discussion happens. However, we want to make the following points:
* The concept is currently sequential only for cubic->ES, all else is parallelizable
* The current impl is sequential per curve, but that's just an impl choice. Personally I'm a fan of discussing this in future work related to work graphs, as I consider load balancing to be a detail
* Parallel lines per ES makes sense but parallel arcs per ES much less so, just because of the prim count. Thus, such detailed discussion may make sense downstream of the prim count results.

Reply:
I'm adding this to the TODO notes.

> Parallel lines per ES makes sense but parallel arcs per ES much less so, just because of the prim count. Thus, such detailed discussion may make sense downstream of the prim count results

Should the prim count discussion happen before or after the GPU Implementation section then? If it's later then this section shouldn't mention it. If it's before then maybe it makes sense to discuss this here.